### ISSUE TITLE
Regarding the training details of the ASR model on limited resources

### ISSUE THREAD

**Issue by msh9184** [NVIDIA/NeMo#6248]
Hi, @titu1994.
I am trying to train a conformer_ctc_bpe (small) model with librispeech960 using four NVIDIA 2080 GPUs (11GB memory).
But, when training model from scratch, I can't reproduce the reported performance with my experimental settings.

Looking at your training details, it seems that a batch size of 2K (maybe 2,048?) and a running rate of 5.0 are used.

However, with my GPUs of 11GB, I can't use such a large batch size; the maximum batch size I can use is 12.

Could you please share some guidelines and tips about appropriate training details for my experimental condition..?
(i.e., **batch size**, **learning rate**, **warmup step**, **weight decay**, **accumulate_grad_batches**, etc. or another optimizer / scheduler ..?)

Some configurations I set are as follows
```
model:
  sample_rate: 16000
  log_prediction: true
  ctc_reduction: 'mean_batch'
  skip_nan_grad: false
  
  train_ds:
    manifest_filepath: "/home/shmun/NeMo/data/train_all_960.json"
    sample_rate: 16000
    batch_size: 12 # you may increase batch_size if your memory allows
    shuffle: true
    num_workers: 8
    pin_memory: true
    use_start_end_token: false
    trim_silence: false
    max_duration: 16.7
    min_duration: 0.1
    is_tarred: false
    tarred_audio_filepaths: null
    shuffle_n: 2048
    bucketing_strategy: "synced_randomized"
    bucketing_batch_size: null

  tokenizer:
    dir: "/home/shmun/NeMo/data/tokenizer_ctc_conformer_large_128/"
    type: "bpe"

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2 
    time_masks: 5  #10
    freq_width: 27
    time_width: 0.05

  optim:
    name: adamw
    lr: 5.0  #2.0
    betas: [0.9, 0.98]
    weight_decay: 1e-3

    sched:
      name: NoamAnnealing
      d_model: 176
      warmup_steps: 10000
      warmup_ratio: null
      min_lr: 1e-6
```

**Comment by titu1994** [comment]
You can use per GPU 8 batch size and do grad accumulation to get to global batch size 512, that should be ok for training. You won't be able to get exactly same score, but it should be in the ballpark. 

The LR and all other hyper parameters used here is for global batch size 2048, so it's possible it will not get similar results with smaller global batch size. 

You can try reducing lr to 2 if you see divergence, if not, 5 is fine. 

Also, with small models, it takes a lot longer to converge. For example it took us roughly 500-1000 epochs of CTC training alone to get to the wer on the model card. 

If CTC is not a requirement, I would suggest use RNNT models, it can get much much faster convergence - usually 100-150 epochs is enough. 

@VahidooX for additional details if I missed anything

**Comment by titu1994** [comment]
If you do decide to use RNNT, small conformer with fused batch size of 1/2/4 and train batch size of 8 will be fine for training. You'll need to experiment a bit with what config of fused batch size prevents oom during training.

**Comment by msh9184** [comment]
@titu1994 many thanks. I'll try :)

**Comment by VahidooX** [comment]
We already have some checkpoints trained on LS here:
https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/stt_en_conformer_ctc_small_ls
https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/stt_en_conformer_transducer_large_ls
https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/stt_en_conformer_ctc_large_ls

Considering your limited resource I recommend the followings:
+ use fp16 instead of fp32 by setting trainer.precision=16. It would help you to double your batch size.
+ Use Hybrid ASR models if you want to train Conformer-CTC models. It would make the convergence of the CTC models more than 3X faster
+ Use fused_batch_size of 1
+ Use FastConformer which uses 2X less memory, it means you can double your batch size. The training would be 2X faster and inference would be 2X faster.

Please use the following config file:
https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/conformer/hybrid_transducer_ctc/conformer_hybrid_transducer_ctc_bpe.yaml

It is for a large model (120M), if you want small model, you need to update it with the table in the config file.

With the following changes, you can make it FastConformer:
+trainer.precision=16
+model.joint.fused_batch_size=1
+model.ecnoder.subsampling_factor=8
+model.encoder.subsampling=dw_striding
+model.encoder.subsampling_conv_channels=256
+model.encoder.conv_kernel_size=9

With this setting, even for a large model, you should be able to fit batch size of 32 on your GPUs. It should work with lr=5 and global batch size of at least 512. Lower global batch sizes may work, but have not tried it. You can give it a shot.

**Comment by msh9184** [comment]
@VahidooX thanks a lot for your comments!

I have some additional questions.
- As I understood, `global batch size` you define here is "`batch_size` * `accumulate_grad_batches` * `devices(number of gpus)`", right? 
In my case, if `a accumulate_grad_batches of 16`, `a batch_size of 8`, and `ddp with 4 gpus` are used, it achieves `global batch size of 512`?
- When reducing the precision, i.e. using fp16 instead of fp32, can I consider explicitly doubling the `global batch size`?

**Comment by VahidooX** [comment]
Yes, that is the definition of global batch size. Usually when you switch to fp16, you can easily double the batch size per GPU. Then you need to make the accumulate_grad_batches in half to keep the same global batch size. accumulate_grad_batches  should be set lower as much as possible as long as you have large enough global batch size.
For example if the largest batch size you can use per GPU is 8 with fp32, then you should be able to use fp16 with batch size of 16 and accumulate_grad_batches of 8 to get to global size of 512. I suggest to try global size of 256 as long as your training does not diverges considering you have just 4 GPUs.

You may take a look at this issue to get more info on how to reproduce our numbers:
https://github.com/NVIDIA/NeMo/issues/3288

Conformer-CTC is around 3X slower to converge to the best WER possible comparing to Transducer models . We trained the CTC models more than 600 epochs while Transducer models converges under 200 easily. With 4 GPUs, it can take some time. You need to use Conformer-Transducer or the Hybrid ASR to converge faster. 

After training, you can use checkpoint averaging to improve the results a little bit.
